Split Raindrops into two files (#2371)
exercism/problem-specifications: exercises/raindrops/instructions.md

# Instructions

Your task is to convert a number into its corresponding raindrop sounds.

If a given number:

> - is divisible by 3, add 'Pling' to the result.
> - is divisible by 5, add 'Plang' to the result.
> - is divisible by 7, add 'Plong' to the result.

Reviewer: I think most languages use double quotes for strings, so maybe we use them there too?

```suggestion
- is divisible by 3, add "Pling" to the result.
- is divisible by 5, add "Plang" to the result.
- is divisible by 7, add "Plong" to the result.
```

> - **is not** divisible by 3, 5, or 7, the result should be the digits of the number.

Reviewer 2: ```suggestion
- **is not** divisible by 3, 5, or 7, the result should be the number as a string.
```

"the digits of the number" sounds like some (ordered) collection of the digits. But the result should definitely be a string.

Reviewer 3: I think that we had decided to be vague on the "type" of the return for this, to allow a language to determine what they need to test for.

However, the results are a ordered collection of the digits when a string as the order must be most significant to the left and least significant digit to the right, and in that order.

Author: > "the digits of the number" sounds like some (ordered) collection of the digits. 

Yeah, I thought this when I was splitting it out, but figured it had been left vague for some reason for languages without strings. But then how do those languages do this exercise at all?

@kotp Do we have any concrete examples of where it's not a string?

Reviewer 3: I do not have a concrete example, but remember talking about how even "Hello World" returning a string in some languages is not "simple".  And so leaving the answer to be a collection is well enough.  It has been some years.

Reviewer 3: "How do they do this exercise at all?"  They might not.  However, having a collection of some kind can store those numeric values, and may also represent the strings as they see fit.  It is the relationship of the numbers being replaced by some value given some criteria that is important for the exercise, not really the strings values, I mean as far as learning value goes, and that does not diminish the "fun" value in my mind (think ciphergraphic kid games).

Author: My sense is that as this exercise is all about strings, we could safely get away with putting "string" in the description for this bit. I think it's much clearer if we can get away with it.

So I think let's fine a concrete counterexample, or go for Glenn's change. And if there's like only one counterexample, maybe that track should just make an exception?

Reviewer 3: Right, just like all the exercises, the tracks are "inspired" by the problem-specifications  data, not ruled by it.

Reviewer 3: If this exercise were a concept exercise, then I would consider that potentially it is all about strings.  In Ruby I usually mentor this as "all about construction and flow control" rather than focusing on the fact that they are strings.  This is what we concentrate on as an Interview question for the simpler version, as well as expansion, when I use this as an interview question.

Reviewer 3: > So I think let's fine a concrete counterexample, or go for Glenn's change. And if there's like only one counterexample, maybe that track should just make an exception?

Not against Glenn's change, just the reason for it is not disqualifying for what is there.  (It is not a contradictory thing.)

Author: Yeah. OK - I'm going to go with it :) Thanks!

## Examples

- 28 is divisible by 7, but not 3 or 5, so the result would be `"Plong"`.
- 30 is divisible by 3 and 5, but not 7, so the result would be `"PlingPlang"`.
- 34 is not divisible by 3, 5, or 7, so the result would be `"34"`.

~~~~exercism/note

> The simplest way to test if one number is evenly divisible by another is to use the [remainder][remainder] or [modulo][modulo] operator.

Author: ```suggestion
A common way to test if one number is evenly divisible by another is to use [the remainder][remainder] or [the modulo operation][modulo].
Most languages provide operators or functions for one or both of these.
```

> [remainder]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Remainder

Author: ```suggestion
[remainder]: https://exercism.org/docs/programming/operators/remainder
```

# Instructions

Your task is to convert a number into a string that contains corresponding raindrop sounds, based on the number's factors.

The rules of `raindrops` are that if a given number:

> - has 3 as a factor, add 'Pling' to the result.

Author: Nice. Updated :)

- is divisible by 5, add 'Plang' to the result.
- is divisible by 7, add 'Plong' to the result.
- **is not** divisible by 3, 5, or 7, the result should be the digits of the number.

## Examples

- 28 is divisible by 7, but not 3 or 5, so the result would be `"Plong"`.
- 30 is divisible by 3 and 5, but not 7, so the result would be `"PlingPlang"`.
- 34 is not divisible by 3, 5, or 7, so the result would be `"34"`.

~~~~exercism/note
The simplest way to test if one number is evenly divisible by another is to use the [remainder][remainder] or [modulo][modulo] operator.

> [remainder]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Remainder

Reviewer 4: You are reffering to javascript here, even if javascript isnt the only language with a remainder operator.

> [remainder]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Remainder

Reviewer 5: You are reffering to javascript here, even if javascript isnt the only language with a remainder operator.

Author: I am, but I think it's a good doc on what the remainder operator is. I'm welcome to alternative links if they're not overly complex (e.g. wikipedia!)

Author: Also, FWIW, most languages have remainder - it's actually modulo that's the rarity.

Reviewer: I do feel that linking to JS docs is a bit odd, but the contents is good. I couldn't find a better page, except for maybe the wiki page.

Reviewer: Another option would be to explain the remainder here

Author: The best solution will be to have a doc on Exercism that we link to. Let's do that (as it's always part of the plan and this is a low-hanging fruit), but let's link this to the JS doc for now until we have that, then update when we've added that doc.

Author: Actually, I'll do it now before we merge.

Author: @ErikSchierboom Updated.

- 34 is not divisible by 3, 5, or 7, so the result would be `"34"`.

~~~~exercism/note
A common way to test if one number is evenly divisible by another is to compare the [remainder][remainder] or [modulus][modulo] to zero.

> Most languages provide operators or functions for one (or both) of these.

Reviewer 3: This looks better, I was going to comment about it being "simplest" as, well, it is gone now, and this is better.
